sw/host/opentitanlib/src/transport/cw310/usb.rs

```diff
@@ -0,0 +1,546 @@
+
+    /// `CMD_FPGAIO_UTIL` is used to configure gpio pins on the SAM3U chip
+    /// which are connected to the FPGA.
+    pub const CMD_FPGAIO_UTIL: u8 = 0x34;
+    /// Requests to the `FPGAIO_UTIL` command.
+    pub const REQ_IO_CONFIG: u16 = 0xA0;
+    pub const REQ_IO_RELEASE: u16 = 0xA1;
+    pub const REQ_IO_OUTPUT: u16 = 0xA2;
+    /// Configuration requests are used in the data packet sent with `REQ_IO_CONFIG`.
+    pub const CONFIG_PIN_INPUT: u8 = 0x01;
+    pub const CONFIG_PIN_OUTPUT: u8 = 0x02;
+    pub const CONFIG_PIN_SPI1_SDO: u8 = 0x10;
+    pub const CONFIG_PIN_SPI1_SDI: u8 = 0x11;
+    pub const CONFIG_PIN_SPI1_SCK: u8 = 0x12;
+    pub const CONFIG_PIN_SPI1_CS: u8 = 0x13;
+
+    /// `CMD_FPGASPI1_XFER` is used to configure and drive SPI transactions
+    /// between the SAM3U chip and the FPGA.
+    pub const CMD_FPGASPI1_XFER: u8 = 0x35;
+    pub const REQ_ENABLE_SPI: u16 = 0xA0;
+    pub const REQ_DISABLE_SPI: u16 = 0xA1;
+    pub const REQ_CS_LOW: u16 = 0xA2;
+    pub const REQ_CS_HIGH: u16 = 0xA3;
+    pub const REQ_SEND_DATA: u16 = 0xA4;
+
+    const LAST_PIN_NUMBER: u8 = 106;
+
+    /// Create a new connection to a CW310 board.
+    pub fn new(
+        usb_vid: Option<u16>,
+        usb_pid: Option<u16>,
+        usb_serial: Option<String>,
+    ) -> Result<Self> {
+        Ok(Backend {
+            usb: UsbBackend::new(usb_vid, usb_pid, usb_serial)?,
+        })
+    }
+
+    /// Gets the usb serial number of the device.
+    pub fn get_serial_number(&self) -> &str {
+        self.usb.get_serial_number()
+    }
+
+    /// Get the firmware build date as a string.
+    pub fn get_firmware_build_date(&self) -> Result<String> {
+        let mut buf = [0u8; 100];
+        let len = self
+            .usb
+            .read_ctrl(Backend::CMD_FW_BUILD_DATE, 0, &mut buf)?;
+        Ok(String::from_utf8_lossy(&buf[0..len]).to_string())
+    }
+
+    /// Get the firmware version.
+    pub fn get_firmware_version(&self) -> Result<[u8; 3]> {
+        let mut buf = [0u8; 3];
+        let _ = self.usb.read_ctrl(Backend::CMD_FW_VERSION, 0, &mut buf)?;
+        Ok(buf)
+    }
+
+    /// Set GPIO `pinname` to either output or input mode.
+    pub fn pin_set_output(&self, pinname: &str, output: bool) -> Result<()> {
+        let pinnum = self.pin_name_to_number(pinname)?;
+        self.usb.send_ctrl(
+            Backend::CMD_FPGAIO_UTIL,
+            Backend::REQ_IO_CONFIG,
+            &[
+                pinnum,
+                if output {
+                    Backend::CONFIG_PIN_OUTPUT
+                } else {
+                    Backend::CONFIG_PIN_INPUT
+                },
+            ],
+        )?;
+        Ok(())
+    }
+
+    /// Get the state of GPIO `pinname`.
+    pub fn pin_get_state(&self, pinname: &str) -> Result<u8> {
+        let pinnum = self.pin_name_to_number(pinname)? as u16;
+        let mut buf = [0u8; 1];
+        self.usb
+            .read_ctrl(Backend::CMD_FPGAIO_UTIL, pinnum, &mut buf)?;
```

Comment:
If you want to be cute, you can use `std::slice::from_mut`:
```rust
let mut state = 0u8;
read_ctrl(slice::from_mut(&mut state))?;
Ok(state)
```

Reply:
[This](https://imgur.com/1XoMPxX) is how you be cute.

While I think I understand the motivation for `slice::from_mut`, there is some tension between providing safety and offending my sense of brevity (and I read the source code for `array::from_mut`.  Not cute on the inside :) )